### ISSUE TITLE
Merge and split incorrectly segmented parts of the records

### ISSUE THREAD

**Issue by MansMeg** [swerik-project/riksdagen-records#18]
In previous work, we have seen how many segments in the record have either been split up incorrectly or merged incorrectly by the OCR engine. We should train a BERT model to split and merge these segmentation errors using a BERT model.

**Comment by MansMeg** [comment]
There is a similar problem for introductions. Now we have introductions that are split in the corpus. These are both marked as introductions so in practice it does not create much problems if the introductions are removed. Although, we should merge these blocks into one introduction block in the long run.

**Comment by MansMeg** [comment]
An idea is to augment data by artificially joining and splitting already existing data.

**Comment by vgkz** [comment]
Results from prototyping models on the same datasets used to train a title prediction model.
# Segment joining
Tried modelling this as a next sentence prediction problem where the labels indicate if they should be joined. 
Currently, only adjacent margins and split titles are labelled as positives. 

Results after training for five epochs with batch size 16, max length 512 and learning rate 3e-5. Note that there are only 31 positive examples in the test set, so the results are on a small sample size. 
|          | Accuracy | Precision | Recall | F<sub>1</sub>    |
| -------- | -------- | --------- | ------ | ------ |
| Train | 0.9996 | 0.9785    | 0.9838    | 0.9811 |
| Validation | 0.9991    | 0.9667    | 1.0    | 0.9831 |
| Test| 0.9991    | 0.9688    | 1.0    | 0.9841 |

# Predicting merged segments
Tried modelling this as a sequence classification problem where the labels indicate whether a sequence is merged or not. I divided this into two different problems: 1. predicting whether a body text is merged with a margin text and 2. predicting whether a title is merged with other body texts. 

## Merged body text and margin

Results after training for five epochs with batch size 16, max length 512 and learning rate 3e-5. Note that there are only 25 merged examples in the test set. 
|            | Accuracy | Precision | Recall | F<sub>1</sub>    |
| ---------- | -------- | --------- | ------ | ------ |
| Train      | 0.9999   | 0.9926    | 1.0    | 0.9963 |
| Validation | 0.9899   | 0.8571    | 0.5714 | 0.6857 |
| Test       | 0.9905   | 0.9375    | 0.6    | 0.7317 |

## Merged title and other text

Results after training for five epochs with batch size 16, max length 512 and learning rate 3e-5. Note that there are only 15 merged examples in the test set. 
|            | Accuracy | Precision | Recall | F<sub>1</sub> |
| ---------- | -------- | --------- | ------ | ------------- |
| Train      | 0.9999   | 0.9970    | 1.0    | 0.9985        |
| Validation | 0.9982   | 0.9565    | 0.9565 | 0.9565        |
| Test       | 0.9983   | 0.9333    | 0.9333 | 0.9333        |

# Segment splitting
Tried modelling this as a token classification problem where each token is either classified as the first token in a segment or not. Token predictions can then be used to split by each token labelled as 1. I've trained and evaluated the model on a subset of the data consisting of segments that should be split assuming that that the models will only be applied to segments that are predicted to be split.

This leaves 667 segments in the training set, 56 in the validation set, and 50 in the test set. The dataset is mostly merged titles and merged margins. 

Results after training for 10 epochs with batch size 16, max length 512 and learning rate 3e-5.

## Evaluation by tokens

|            | Accuracy | Precision | Recall | F<sub>1</sub> |
| ---------- | -------- | --------- | ------ | ------------- |
| Train         | 0.9994   | 0.9937    | 0.9937 | 0.9937       |
| Validation | 0.9926   | 0.8804    | 0.8617 | 0.8710        |
| Test          | 0.9793   | 0.8289    | 0.75     | 0.7875        |

## Evaluation by segments

|            | Accuracy |
| ---------- | -------- |
| Train         | 0.9730  |
| Validation | 0.75      |
| Test          | 0.6        |

**Comment by MansMeg** [comment]
Great work, @vgkz! The segment joining is at a good enough level to be used, I think, but the others are still too shaky.

@BobBorges @fredrik1984 @ninpnin ?

**Comment by ninpnin** [comment]
IMO we could try all the models and see how they perform on the real data. But I agree with you Måns, the segment joining is looking promising, while the rest might be too unreliable.

**Comment by vgkz** [comment]
I've made a PR #41 with a diff sample from running the joining model on the corpus.
